crates/communication/src/server/receiver.rs

```diff
@@ -0,0 +1,143 @@
+use futures_util::{stream::SplitStream, StreamExt};
+use serde_json::from_str;
+use tokio::{net::TcpStream, select, sync::mpsc::Sender};
+use tokio_tungstenite::{
+    tungstenite::{protocol::frame::coding::CloseCode, Message},
+    WebSocketStream,
+};
+use tokio_util::sync::CancellationToken;
+
+use crate::server::{
+    databases::{Client, ClientRequest},
+    messages::DatabaseRequest,
+};
+
+use super::{
+    connection::ReceiverOrSenderError,
+    databases,
+    messages::{Request, Response},
+};
+
+pub async fn receiver(
+    mut reader: SplitStream<WebSocketStream<TcpStream>>,
+    error_sender: Sender<ReceiverOrSenderError>,
+    keep_running: CancellationToken,
+    keep_only_self_running: CancellationToken,
+    client_id: usize,
+    response_sender: Sender<Response>,
+    databases_sender: Sender<databases::Request>,
+) {
+    select! {
+        _ = async {
+            while let Some(message) = reader.next().await {
+                handle_message(
+                    message,
+                    &error_sender,
+                    &keep_only_self_running,
+                    client_id,
+                    &response_sender,
+                    &databases_sender,
+                ).await;
+            }
+        } => {},
+        _ = keep_running.cancelled() => {},
+        _ = keep_only_self_running.cancelled() => {},
+    }
+
+    let _ = databases_sender
+        .send(databases::Request::ClientRequest(ClientRequest {
+            request: DatabaseRequest::UnsubscribeEverything,
+            client: Client {
+                id: client_id,
+                response_sender: response_sender.clone(),
+            },
+        }))
+        .await;
```

Comment:
The `connection()` function only lives during construction of the `sender()` and `receiver()` tasks. It would not be possible to do it there. IMHO, it makes sense to have this in the receiver because we kinda inject a message to the router and pretend it came from the real client.